contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -291,54 +371,69 @@ fn apply_transfer(
     source_chain: ChainNameRaw,
     destination_chain: ChainNameRaw,
     transfer: &InterchainTransfer,
-) -> Result<(), Error> {
-    subtract_amount_from_source(storage, source_chain, transfer)?;
-    add_amount_to_destination(storage, destination_chain, transfer)
+) -> Result<InterchainTransfer, Error> {
+    let destination_amount = destination_amount(
+        storage,
+        &source_chain,
+        &destination_chain,
+        transfer.token_id,
+        transfer.amount,
+    )?;
+
+    subtract_amount_from_source(storage, source_chain, transfer.token_id, transfer.amount)?;
+    add_amount_to_destination(
+        storage,
+        destination_chain,
+        transfer.token_id,
+        destination_amount,
+    )?;
+
+    Ok(InterchainTransfer {
+        amount: destination_amount,
+
+        ..transfer.clone()
+    })
```

Comment:
I would continue passing in the `InterchainTransfer` object to `subtract_amount_from_source` and `add_amount_to_destination`, and just transform the amount in the object in between those two calls.